internal/dnfjson/dnfjson.go

```diff
@@ -0,0 +1,347 @@
+package dnfjson
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/osbuild/osbuild-composer/internal/rhsm"
+	"github.com/osbuild/osbuild-composer/internal/rpmmd"
+)
+
+// Solver is configured with system information in order to resolve
+// dependencies for RPM packages using DNF.
+type Solver struct {
+	// Platform ID, e.g., "platform:el8"
+	ModulePlatformID string `json:"module_platform_id"`
+
+	// System architecture
+	Arch string `json:"arch"`
+
+	// Cache directory for the DNF metadata
+	CacheDir string `json:"cachedir"`
+
+	// Release version of the distro. This is used in repo files on the host
+	// system and required for subscription support.
+	releaseVer string `json:"-"`
+
+	// Path to the dnf-json binary and optional args (default: "osbuild-dnf-json", assumed in $PATH)
+	dnfJsonCmd []string `json:"-"`
+}
+
+// Create a new unconfigured Solver (without platform information). It should
+// be configured using the SetConfig() function before use.
+func NewBaseSolver(cacheDir string) *Solver {
+	return &Solver{
+		CacheDir:   cacheDir,
+		dnfJsonCmd: []string{"osbuild-dnf-json"},
+	}
+}
+
+// Create a new Solver with the given configuration
+func NewSolver(modulePlatformID string, releaseVer string, arch string, cacheDir string) *Solver {
+	return &Solver{
+		ModulePlatformID: modulePlatformID,
+		Arch:             arch,
+		CacheDir:         cacheDir,
+		releaseVer:       releaseVer,
+		dnfJsonCmd:       []string{"osbuild-dnf-json"},
+	}
+}
+
+// SetConfig sets the platform configuration values.
+func (s *Solver) SetConfig(modulePlatformID string, releaseVer string, arch string) {
+	s.ModulePlatformID = modulePlatformID
+	s.Arch = arch
+	s.releaseVer = releaseVer
+}
```

Comment:
It's not uncommon in the project to call `Depsolve()` multiple times with the same platform configuration.  I find it's convenient to have a solver and reuse it.
For example, I find this is looks cleaner than having 6 arguments to the `Depsolve()` call:
https://github.com/osbuild/osbuild-composer/blob/bfca5d5dac89eed13b9497bd3d93b2d02f2cf9cd/cmd/osbuild-worker/jobimpl-depsolve.go#L31

Having a `BaseDepsolver` and a `SetConfig()` lets us carry around a Solver config with any information that we know wont change: cache directory and subscriptions (though this isn't in yet).  Maybe we could also attach the architecture to the base config as well.

Similarly, the `rpmmd` instance we keep on API instances does the same.

We could also initialise multiple solvers, one per supported distro to reuse and carry all around for when we need them, but that might be a bit much.

Reply:
I see your point, but I tend to disagree with this approach. I don't find it less convenient to pass more variables to the function, especially when these tend to be potentially set to new values before each API call handling.

The part that I don't like is that the `Solver` instance is basically left tainted after calling `SetConfig()` and one can not rely on its state before using it again for depsolving. 

Since the same `Solver` instance is used in composer for both Weldr and Koji APIs, I'm afraid that in a very rare and unlikely case, when Weldr and Koji APIs are enabled and used on the same composer instance, the changes of the `Solver` instance in one API code could conflict with depsolving being done in the other API code.

A solution that I would find more acceptable could be something like:

1. Instantiate a `Solver` when starting up composer, with the `ModulePlatformID`, `Arch` and `ReleaseVer`, based on the OS on which composer is running.
2. Where relevant, have a variant of `SetConfig()` which would return a clone of the original `Solver` with specific `ModulePlatformID`, `Arch` and `ReleaseVer` values set in it. This instance would be thrown away after handling an API call.

Reply:
> I see your point, but I tend to disagree with this approach. I don't find it less convenient to pass more variables to the function, especially when these tend to be potentially set to new values before each API call handling.
> 
> The part that I don't like is that the `Solver` instance is basically left tainted after calling `SetConfig()` and one can not rely on its state before using it again for depsolving.

This is a great point that could lead to issues.

> Since the same `Solver` instance is used in composer for both Weldr and Koji APIs, I'm afraid that in a very rare and unlikely case, when Weldr and Koji APIs are enabled and used on the same composer instance, the changes of the `Solver` instance in one API code could conflict with depsolving being done in the other API code.
> 
> A solution that I would find more acceptable could be something like:
> 
> 1. Instantiate a `Solver` when starting up composer, with the `ModulePlatformID`, `Arch` and `ReleaseVer`, based on the OS on which composer is running.
> 2. Where relevant, have a variant of `SetConfig()` which would return a clone of the original `Solver` with specific `ModulePlatformID`, `Arch` and `ReleaseVer` values set in it. This instance would be thrown away after handling an API call.

I like this.  Could also be used with uninitialised `ModulePlatformID`, `Arch` and `ReleaseVer`, so that the Solver only holds the cache dir and the subscriptions (unless we make a point of reloading them every time).

I'll do this.

Reply:
> It's not uncommon in the project to call Depsolve() multiple times with the same platform configuration.

As discussed offline, it might be worth it eventually to avoid multiple `Depsolve()` calls, but rather allow it to take an array of arguments. This would mostly be a potential optimisation as you would only need to call `dnf-json` once, and could perhaps avoid some overhead, but also we could make a guarantee that several calls would not get the caches invalidated (i.e., the buildroot and the payload would have the same exact package versions, for instance).